File: backend/cmd/server/sync.go
package main

import (
	"encoding/json"
	"net/http"
	"time"

	"github.com/labstack/echo/v5"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/apis"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/models"
)

// SyncPushRequest represents a sync push request
type SyncPushRequest struct {
	TenantID        string       `json:"tenantId"`
	Changes         []SyncChange `json:"changes"`
	ClientTimestamp string       `json:"clientTimestamp"`
}

// SyncChange represents a single change to sync
type SyncChange struct {
	EntityType   string                 `json:"entityType"`
	EntityID     string                 `json:"entityId"`
	Operation    string                 `json:"operation"` // create, update, delete
	Data         map[string]interface{} `json:"data"`
	LocalVersion int                    `json:"localVersion"`
}

// SyncPullResponse represents a sync pull response
type SyncPullResponse struct {
	Changes         []SyncChange `json:"changes"`
	ServerTimestamp string       `json:"serverTimestamp"`
	HasMore         bool         `json:"hasMore"`
}

// SyncConflict represents a sync conflict
type SyncConflict struct {
	EntityType string                 `json:"entityType"`
	EntityID   string                 `json:"entityId"`
	LocalData  map[string]interface{} `json:"localData"`
	RemoteData map[string]interface{} `json:"remoteData"`
}

// SyncResult represents the result of a sync operation
type SyncResult struct {
	Success   bool           `json:"success"`
	Synced    int            `json:"synced"`
	Conflicts []SyncConflict `json:"conflicts,omitempty"`
	Errors    []string       `json:"errors,omitempty"`
}

// setupSyncRoutes configures the sync API endpoints
func setupSyncRoutes(e *core.ServeEvent, app *pocketbase.PocketBase) {
	// Push local changes to server
	e.Router.POST("/api/sync/push", func(c echo.Context) error {
		return handleSyncPush(c, app)
	}, apis.RequireRecordAuth())

	// Pull remote changes from server
	e.Router.GET("/api/sync/pull", func(c echo.Context) error {
		return handleSyncPull(c, app)
	}, apis.RequireRecordAuth())

	// Resolve a conflict
	e.Router.POST("/api/sync/resolve", func(c echo.Context) error {
		return handleSyncResolve(c, app)
	}, apis.RequireRecordAuth())

	// Get sync status
	e.Router.GET("/api/sync/status", func(c echo.Context) error {
		return handleSyncStatus(c, app)
	}, apis.RequireRecordAuth())
}

// handleSyncPush processes incoming changes from client
func handleSyncPush(c echo.Context, app *pocketbase.PocketBase) error {
	var req SyncPushRequest
	if err := c.Bind(&req); err != nil {
		return apis.NewBadRequestError("Invalid request body", err)
	}

	// Validate tenant access
	if err := validateTenantAccess(c, req.TenantID); err != nil {
		return apis.NewForbiddenError(err.Error(), nil)
	}

	result := SyncResult{
		Success: true,
		Synced:  0,
	}

	for _, change := range req.Changes {
		err := processChange(app, req.TenantID, change)
		if err != nil {
			result.Errors = append(result.Errors, err.Error())
			result.Success = false
		} else {
			result.Synced++
		}
	}

	return c.JSON(http.StatusOK, result)
}

// handleSyncPull retrieves changes since a given timestamp
func handleSyncPull(c echo.Context, app *pocketbase.PocketBase) error {
	tenantID := getTenantID(c)
	if tenantID == "" {
		return apis.NewBadRequestError("Tenant ID is required", nil)
	}

	since := c.QueryParam("since")
	if since == "" {
		// Default to 24 hours ago
		since = time.Now().Add(-24 * time.Hour).Format(time.RFC3339)
	}

	// Get changes from various collections
	changes := []SyncChange{}
	collections := []string{"products", "categories", "customers", "sales", "inventory_movements"}

	for _, collName := range collections {
		coll, err := app.Dao().FindCollectionByNameOrId(collName)
		if err != nil {
			continue
		}

		records, err := app.Dao().FindRecordsByFilter(
			coll.Id,
			"tenant_id = {:tenantId} && updated >= {:since}",
			"-updated",
			100,
			0,
			map[string]interface{}{
				"tenantId": tenantID,
				"since":    since,
			},
		)
		if err != nil {
			continue
		}

		for _, record := range records {
			data := map[string]interface{}{}
			for key, value := range record.PublicExport() {
				data[key] = value
			}
			changes = append(changes, SyncChange{
				EntityType:   collName,
				EntityID:     record.Id,
				Operation:    "update",
				Data:         data,
				LocalVersion: record.GetInt("sync_version"),
			})
		}
	}

	response := SyncPullResponse{
		Changes:         changes,
		ServerTimestamp: time.Now().Format(time.RFC3339),
		HasMore:         false, // Implement pagination if needed
	}

	return c.JSON(http.StatusOK, response)
}

// handleSyncResolve processes conflict resolution
func handleSyncResolve(c echo.Context, app *pocketbase.PocketBase) error {
	var req struct {
		ConflictID string                 `json:"conflictId"`
		Resolution string                 `json:"resolution"` // local_wins, remote_wins, merged
		MergedData map[string]interface{} `json:"mergedData,omitempty"`
	}

	if err := c.Bind(&req); err != nil {
		return apis.NewBadRequestError("Invalid request body", err)
	}

	// Find the conflict record
	collection, err := app.Dao().FindCollectionByNameOrId("sync_conflicts")
	if err != nil {
		return apis.NewNotFoundError("Sync conflicts collection not found", err)
	}

	conflict, err := app.Dao().FindRecordById(collection.Id, req.ConflictID)
	if err != nil {
		return apis.NewNotFoundError("Conflict not found", err)
	}

	// Apply resolution
	conflict.Set("resolution", req.Resolution)
	conflict.Set("resolved_at", time.Now().Format(time.RFC3339))

	if req.Resolution == "merged" && req.MergedData != nil {
		mergedJSON, _ := json.Marshal(req.MergedData)
		conflict.Set("merged_data", string(mergedJSON))
	}

	if err := app.Dao().SaveRecord(conflict); err != nil {
		return apis.NewBadRequestError("Failed to save resolution", err)
	}

	return c.JSON(http.StatusOK, map[string]string{
		"status": "resolved",
	})
}

// handleSyncStatus returns current sync status for the tenant
func handleSyncStatus(c echo.Context, app *pocketbase.PocketBase) error {
	tenantID := getTenantID(c)
	if tenantID == "" {
		return apis.NewBadRequestError("Tenant ID is required", nil)
	}

	// Count pending conflicts
	conflictsCollection, _ := app.Dao().FindCollectionByNameOrId("sync_conflicts")
	pendingConflicts := 0
	if conflictsCollection != nil {
		records, _ := app.Dao().FindRecordsByFilter(
			conflictsCollection.Id,
			"tenant_id = {:tenantId} && resolution = ''",
			"",
			0,
			0,
			map[string]interface{}{"tenantId": tenantID},
		)
		pendingConflicts = len(records)
	}

	return c.JSON(http.StatusOK, map[string]interface{}{
		"tenantId":         tenantID,
		"pendingConflicts": pendingConflicts,
		"serverTime":       time.Now().Format(time.RFC3339),
	})
}

// processChange applies a single change to the database
func processChange(app *pocketbase.PocketBase, tenantID string, change SyncChange) error {
	collection, err := app.Dao().FindCollectionByNameOrId(change.EntityType)
	if err != nil {
		return err
	}

	switch change.Operation {
	case "create":
		record := models.NewRecord(collection)
		for key, value := range change.Data {
			record.Set(key, value)
		}
		record.Set("tenant_id", tenantID)
		record.Set("sync_status", "synced")
		record.Set("sync_version", change.LocalVersion+1)
		return app.Dao().SaveRecord(record)

	case "update":
		record, err := app.Dao().FindRecordById(collection.Id, change.EntityID)
		if err != nil {
			return err
		}

		// Check for conflicts
		serverVersion := record.GetInt("sync_version")
		if serverVersion > change.LocalVersion {
			// Conflict detected - could create a conflict record here
			return nil // For now, skip conflicting updates
		}

		for key, value := range change.Data {
			if key != "id" && key != "tenant_id" {
				record.Set(key, value)
			}
		}
		record.Set("sync_status", "synced")
		record.Set("sync_version", serverVersion+1)
		return app.Dao().SaveRecord(record)

	case "delete":
		record, err := app.Dao().FindRecordById(collection.Id, change.EntityID)
		if err != nil {
			return err
		}
		return app.Dao().DeleteRecord(record)
	}

	return nil
}

File: packages/server/src/routes/sync.ts
/**
 * Sync Routes
 *
 * Handles local sync queue management and external sync operations.
 *
 * LOCAL QUEUE (Implemented - Phase 1):
 * - POST   /api/sync/queue          - Add operation to local sync queue
 * - GET    /api/sync/queue          - Get pending operations
 * - DELETE /api/sync/queue/:id      - Remove operation from queue
 * - GET    /api/sync/status         - Get sync status
 *
 * EXTERNAL SYNC (Returns 501 - Phase 2):
 * - POST   /api/sync/push           - Push local changes to remote server
 * - GET    /api/sync/pull           - Pull remote changes
 * - POST   /api/sync/resolve        - Resolve sync conflict
 *
 * @module routes/sync
 *
 * ============================================================================
 * PHASE 2 IMPLEMENTATION NOTES
 * ============================================================================
 *
 * When implementing external sync, the following endpoints need to be completed:
 *
 * 1. POST /api/sync/push
 *    - Accept array of sync queue items
 *    - For each item:
 *      a. Send to remote server
 *      b. Handle conflicts (version mismatch)
 *      c. Update local sync_status to 'synced' or 'conflict'
 *      d. Remove from sync_queue on success
 *    - Request body: { changes: SyncChange[], clientTimestamp: string }
 *    - Response: { success: boolean, synced: number, conflicts: SyncConflict[] }
 *
 * 2. GET /api/sync/pull
 *    - Accept ?since=timestamp parameter
 *    - Fetch changes from remote server since that timestamp
 *    - Apply changes to local database
 *    - Handle conflicts with local pending changes
 *    - Response: { changes: SyncChange[], serverTimestamp: string, hasMore: boolean }
 *
 * 3. POST /api/sync/resolve
 *    - Accept conflict resolution: { conflictId: string, resolution: 'local_wins' | 'remote_wins' | 'merged', mergedData?: object }
 *    - Apply resolution to sync_conflicts table
 *    - Update the entity with resolved data
 *    - Response: { success: boolean, entity: object }
 *
 * Configuration needed for Phase 2:
 * - Remote server URL (e.g., SYNC_SERVER_URL env var)
 * - Authentication token for remote server
 * - Sync interval (for background sync)
 * - Retry policy (exponential backoff)
 *
 * Suggested libraries:
 * - node-fetch or undici for HTTP requests
 * - p-retry for retry logic
 * - cron or node-schedule for background sync
 *
 * ============================================================================
 */

import { FastifyInstance, FastifyRequest, FastifyReply } from 'fastify';
import { eq, and, desc } from 'drizzle-orm';
import { nanoid } from 'nanoid';
import { syncQueue, syncConflicts } from '../db/schema.js';

interface TokenPayload {
  userId: string;
  tenantId: string;
  email: string;
  role: string;
}

interface SyncQueueBody {
  entityType: string;
  entityId: string;
  operation: 'create' | 'update' | 'delete';
  data?: Record<string, unknown>;
}

export async function syncRoutes(app: FastifyInstance): Promise<void> {
  // Authentication hook for all sync routes
  app.addHook('preHandler', async (request: FastifyRequest, reply: FastifyReply) => {
    try {
      await request.jwtVerify();
      const payload = request.user as TokenPayload;
      request.tenantId = payload.tenantId;
    } catch (err) {
      reply.status(401).send({ error: 'Unauthorized', message: 'Invalid or expired token' });
    }
  });

  // ============================================================================
  // LOCAL SYNC QUEUE ENDPOINTS (Phase 1 - Implemented)
  // ============================================================================

  /**
   * POST /api/sync/queue
   * Add an operation to the local sync queue
   */
  app.post<{ Body: SyncQueueBody }>('/queue', {
    schema: {
      body: {
        type: 'object',
        required: ['entityType', 'entityId', 'operation'],
        properties: {
          entityType: { type: 'string' },
          entityId: { type: 'string' },
          operation: { type: 'string', enum: ['create', 'update', 'delete'] },
          data: { type: 'object' },
        },
      },
    },
    handler: async (request, reply) => {
      const { entityType, entityId, operation, data } = request.body;
      const tenantId = request.tenantId;

      if (!tenantId) {
        return reply.status(400).send({ error: 'Tenant ID required' });
      }

      const now = new Date().toISOString();
      const id = nanoid();

      await app.db.insert(syncQueue).values({
        id,
        tenantId,
        entityType,
        entityId,
        operation,
        data: data || {},
        localVersion: 1,
        attempts: 0,
        createdAt: now,
      });

      return reply.status(201).send({
        id,
        entityType,
        entityId,
        operation,
        createdAt: now,
      });
    },
  });

  /**
   * GET /api/sync/queue
   * Get pending operations from the sync queue
   */
  app.get<{ Querystring: { limit?: string } }>('/queue', async (request, reply) => {
    const tenantId = request.tenantId;
    const limit = Math.min(100, parseInt(request.query.limit || '50', 10));

    if (!tenantId) {
      return reply.status(400).send({ error: 'Tenant ID required' });
    }

    const items = await app.db
      .select()
      .from(syncQueue)
      .where(eq(syncQueue.tenantId, tenantId))
      .orderBy(syncQueue.createdAt)
      .limit(limit)
      .all();

    return {
      items,
      count: items.length,
    };
  });

  /**
   * DELETE /api/sync/queue/:id
   * Remove an operation from the sync queue (after successful sync)
   */
  app.delete<{ Params: { id: string } }>('/queue/:id', async (request, reply) => {
    const { id } = request.params;
    const tenantId = request.tenantId;

    if (!tenantId) {
      return reply.status(400).send({ error: 'Tenant ID required' });
    }

    // Verify the item belongs to the tenant
    const item = await app.db
      .select()
      .from(syncQueue)
      .where(and(eq(syncQueue.id, id), eq(syncQueue.tenantId, tenantId)))
      .get();

    if (!item) {
      return reply.status(404).send({ error: 'Sync queue item not found' });
    }

    await app.db.delete(syncQueue).where(eq(syncQueue.id, id));

    return { success: true, id };
  });

  /**
   * GET /api/sync/status
   * Get the current sync status
   */
  app.get('/status', async (request, reply) => {
    const tenantId = request.tenantId;

    if (!tenantId) {
      return reply.status(400).send({ error: 'Tenant ID required' });
    }

    // Count pending queue items
    const queueItems = await app.db
      .select()
      .from(syncQueue)
      .where(eq(syncQueue.tenantId, tenantId))
      .all();

    // Count unresolved conflicts
    const conflicts = await app.db
      .select()
      .from(syncConflicts)
      .where(and(eq(syncConflicts.tenantId, tenantId), eq(syncConflicts.status, 'pending')))
      .all();

    return {
      pendingCount: queueItems.length,
      conflictsCount: conflicts.length,
      externalSyncEnabled: false, // Phase 2: Change to true when implemented
      lastSyncAt: null, // Phase 2: Store and return last sync timestamp
      status: queueItems.length === 0 ? 'synced' : 'pending',
    };
  });

  /**
   * GET /api/sync/conflicts
   * Get unresolved sync conflicts
   */
  app.get('/conflicts', async (request, reply) => {
    const tenantId = request.tenantId;

    if (!tenantId) {
      return reply.status(400).send({ error: 'Tenant ID required' });
    }

    const conflicts = await app.db
      .select()
      .from(syncConflicts)
      .where(and(eq(syncConflicts.tenantId, tenantId), eq(syncConflicts.status, 'pending')))
      .orderBy(desc(syncConflicts.createdAt))
      .all();

    return {
      items: conflicts,
      count: conflicts.length,
    };
  });

  // ============================================================================
  // EXTERNAL SYNC ENDPOINTS (Phase 2 - Returns 501 Not Implemented)
  // ============================================================================

  /**
   * POST /api/sync/push
   * Push local changes to remote server
   *
   * PHASE 2 TODO:
   * - Accept SyncPushRequest with changes array
   * - Send each change to remote server
   * - Handle version conflicts
   * - Update local sync_status
   * - Remove successfully synced items from queue
   */
  app.post('/push', async (_request, reply) => {
    return reply.status(501).send({
      error: 'Not Implemented',
      message: 'External sync push will be available in Phase 2',
      phase: 2,
      endpoint: 'POST /api/sync/push',
      expectedRequest: {
        changes: [
          {
            entityType: 'products',
            entityId: 'abc123',
            operation: 'create|update|delete',
            data: {},
            localVersion: 1,
          },
        ],
        clientTimestamp: '2024-01-01T00:00:00.000Z',
      },
      expectedResponse: {
        success: true,
        synced: 5,
        conflicts: [],
        errors: [],
      },
    });
  });

  /**
   * GET /api/sync/pull
   * Pull remote changes from server
   *
   * PHASE 2 TODO:
   * - Accept ?since=timestamp query parameter
   * - Fetch changes from remote server
   * - Apply changes to local database
   * - Detect conflicts with local pending changes
   */
  app.get('/pull', async (_request, reply) => {
    return reply.status(501).send({
      error: 'Not Implemented',
      message: 'External sync pull will be available in Phase 2',
      phase: 2,
      endpoint: 'GET /api/sync/pull?since=timestamp',
      expectedResponse: {
        changes: [
          {
            entityType: 'products',
            entityId: 'abc123',
            operation: 'create|update|delete',
            data: {},
            remoteVersion: 2,
          },
        ],
        serverTimestamp: '2024-01-01T00:00:00.000Z',
        hasMore: false,
      },
    });
  });

  /**
   * POST /api/sync/resolve
   * Resolve a sync conflict
   *
   * PHASE 2 TODO:
   * - Accept conflict ID and resolution strategy
   * - Apply resolution (local_wins, remote_wins, or merged data)
   * - Update entity with resolved data
   * - Mark conflict as resolved
   */
  app.post('/resolve', async (_request, reply) => {
    return reply.status(501).send({
      error: 'Not Implemented',
      message: 'Conflict resolution will be available in Phase 2',
      phase: 2,
      endpoint: 'POST /api/sync/resolve',
      expectedRequest: {
        conflictId: 'conflict_abc123',
        resolution: 'local_wins|remote_wins|merged',
        mergedData: {},
      },
      expectedResponse: {
        success: true,
        entity: {},
        resolution: 'local_wins',
      },
    });
  });
}
